File: src/dotnet/EntityFX.NetBenchamarks.Core/Writer.cs
using System;
using System.IO;

namespace EntityFX.NetBenchmark.Core
{
    public class Writer : IDisposable, EntityFX.NetBenchmark.Core.IWriter
    {
        private TextWriter writer = new StringWriter();

        private StreamWriter streamWriter;

        public bool UseConsole { get; set; }

        public bool UseFile { get; set; }

        public string FilePath { get; private set; }

        public Writer(string filePath)
        {
            UseConsole = true;
            if (filePath != null)
            {
                FilePath = filePath;
                UseFile = true;

                FileStream mystream = new FileStream(filePath,
                    FileMode.OpenOrCreate, FileAccess.Write);

                streamWriter = new StreamWriter(mystream);
                streamWriter.AutoFlush = true;
            }
        }

        public string Output {
            get { 
                return writer.ToString(); 
            }
        }

        public void WriteLine(string format, params object[] args)
        {
            Write(ConsoleColor.Gray, format, args);
            WriteLine();
        }

        public void WriteLine()
        {
            writer.WriteLine();
            if (UseConsole) Console.WriteLine();
            if (UseFile)
            {
                streamWriter.WriteLine();
            }
        }

        public void WriteHeader(string format, params object[] args)
        {
            writer.WriteLine(format, args);
            Write(ConsoleColor.Cyan, format, args);
            WriteLine();
        }

        public void Write(string format, params object[] args)
        {
            Write(ConsoleColor.Gray, format, args);
        }

        public void Write(ConsoleColor color, string format, params object[] args)
        {
            writer.Write(format, args);
            if (UseConsole)
            {
#if NETSTANDARD2_0
                var tmpColor = Console.ForegroundColor;
                Console.ForegroundColor = color;
#endif
                Console.Write(format, args);
#if NETSTANDARD2_0
                Console.ForegroundColor = tmpColor;
#endif
            }
            if (UseFile)
            {
                streamWriter.Write(format, args);
            }
        }

        public void WriteValue(string format, params object[] args)
        {
            Write(ConsoleColor.Green, format, args);
        }


        public void WriteTitle(string format, params object[] args)
        {
            Write(ConsoleColor.White, format, args);
        }

        public void Dispose()
        {
            streamWriter.Close();
            streamWriter.Dispose();
        }
    }
}

File: src/java/EntityFX/src/EntityFX/Core/Writer.java
package EntityFX.Core;

import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;

public class Writer {

    private final String writer = "";

    public boolean UseConsole = true;

    public String FilePath = null;

    public String Output;

    private FileOutputStream fileStream;

    public Writer(final String filePath) throws FileNotFoundException {
        if (filePath != null) {
            this.FilePath = filePath;
            this.fileStream = new FileOutputStream(filePath, true);
        }
    }

    public void writeLine(final String format, Object... args) throws IOException {
        this.writeColor("\033[1;30m", format, args);
        this.writeLine();
    }

    public void writeLine() throws IOException
    {
        if (this.UseConsole) {
            if (this.FilePath != null) {
                fileStream.write("\n".getBytes());
            }
            System.out.println();
        }
        this.Output += "\n";
    }

    public void writeHeader(String format, Object... args) throws IOException
    {
        this.writeColor("\033[1;36m", format, args);
        this.writeLine();
    }

    public void write(String format, Object... args) throws IOException
    {
        this.writeColor("\033[1;30m", format, args);
    }

    public void writeColor(final String color, String format, Object... args) throws IOException
    {
        var formatted = String.format(format, args);
        if (this.UseConsole)
        {
            System.out.print(color + formatted + "\033[0m");

            if (this.FilePath != null) {
                fileStream.write(formatted.getBytes());
            }
        }
        this.Output += formatted;
    }

    public void writeValue(String format, Object... args) throws IOException
    {
        this.writeColor("\033[1;32m", format, args);
    }

    public void writeTitle(String format, Object... args) throws IOException
    {
        this.writeColor("\033[1;37m", format, args);
    }
    
}